apstools/devices/tests/test_motor_factory.py

```diff
@@ -0,0 +1,279 @@
+"""Test the motor bundle factories."""
+
+from contextlib import nullcontext as does_not_raise
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Type
+from typing import Union
+
+import pytest
+from ophyd import Component
+from ophyd import Device
+from ophyd import MotorBundle
+from ophyd import ophydobj
+
+from ..motor_factory import axis_component
+from ..motor_factory import mb_class_factory
+from ..motor_factory import mb_creator
+from ..motor_factory import MOTORS_TYPE
+
+
+class Mixin(Device):
+    """Device class for testing."""
+
+    _mixin = True
+
+
+@pytest.mark.parametrize(
+    "parms, class_str, raises, expected",
+    [
+        [None, "SoftPositioner", None, None],  # no PV, default to soft
+        [dict(), "SoftPositioner", None, None],  # same
+        ["m1", "EpicsMotor", None, None],  # PV str, use a motor
+        [dict(prefix="m1"), "EpicsMotor", None, None],  # same
+        [  # dict with a pv ('prefix', by name)
+            dict(
+                prefix="m1",
+                labels=["dcm"],
+            ),
+            "EpicsMotor",
+            None,
+            None,
+        ],
+        [  # can't specify a class AND a factory
+            {
+                "class": "abcde",
+                "factory": "factory",
+            },
+            None,
+            ValueError,
+            "Unsupported configuration",
+        ],
+        [  # specify a class
+            {
+                "class": "apstools.devices.PVPositionerSoftDoneWithStop",
+                "readback_pv": "gp:float14",
+                "setpoint_pv": "gp:float4",
+            },
+            "PVPositionerSoftDoneWithStop",
+            None,
+            None,
+        ],
+        [  # specify a class factory
+            {
+                "factory": {
+                    "function": "apstools.devices.mb_class_factory",
+                    "motors": ["x1", "x2", "y1", "y2", "y3", "z"],
+                    "class_name": "TestDevice",
+                },
+            },
+            "TestDevice",
+            None,
+            None,
+        ],
+    ],
+)
+def test_axis_component(
+    parms: Dict[str, Any], 
```

Comment:
Parms can also be None (line 31) or str (line 33).

Reply:
Copied `parms` typing from `motor_factory.py`